### PR TITLE
Refactor RewardTrainer hyperparameters into dedicated dataclass

### PR DESCRIPTION
**Pull request by lewtun**
This PR migrates the `max_length` arg of the `RewardTrainer` into a dedicated `RewardTrainingArguments` class that can also be used for storing future hyperparameters.

To be backwards compatible, I've left the variable in the trainer's init, with a warning that this will be removed in some future version.

Tested with:

```shell
accelerate launch --multi_gpu --num_processes 2 examples/scripts/reward_trainer.py --batch_size 1
```

### CODE REVIEW

File: trl/trainer/reward_trainer.py [hunk 1]
```diff
@@ -94,18 +95,21 @@ def __init__(
                 The optimizer and scheduler to use for training.
             preprocess_logits_for_metrics (`Callable[[torch.Tensor, torch.Tensor], torch.Tensor]`):
                 The function to use to preprocess the logits before computing the metrics.
-            max_length (`int`, defaults to `None`):
```
Comment (lewtun) [inline trl/trainer/reward_trainer.py:97]: I followed the `transformers` convention to remove deprecated args from public doc strings

File: trl/trainer/reward_trainer.py [hunk 2]
```diff
@@ -94,11 +95,14 @@ def __init__(
...
-                The maximum length of the sequences in the batch. This argument is required if you want to use the default data collator.
             peft_config (`Dict`, defaults to `None`):
                 The PEFT configuration to use for training. If you pass a PEFT configuration, the model will be wrapped in a PEFT model.
         """
+        if max_length is not None:
+            warnings.warn(
+                "The `max_length` argument is deprecated and will be removed in a future version. Please use the `RewardTrainingArguments` to set `max_length` instead.",
```
Comment (lewtun) [inline trl/trainer/reward_trainer.py:103]: Should we specify a precise version, e.g. 0.9.0?

File: trl/trainer/training_args.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from dataclasses import dataclass, field
+from typing import Optional
+
+from transformers import TrainingArguments
+
+
+@dataclass
+class RewardTrainingArguments(TrainingArguments):
```
Comment (lewtun) [inline trl/trainer/training_args.py:23]: Another name for this could be `RewardConfig` to be more aligned with `PPOConfig`

Comment (lewtun) [reply trl/trainer/training_args.py]: I opted for `RewardConfig` in https://github.com/huggingface/trl/pull/726/commits/e6e346d321654de7a689d44a05ef51d9e2005a34

I've also placed this in a `training_configs.py` module - let me know if you'd prefer a dedicated module per config (currently `PPOConfig` lives in `ppo_config.py`)

**Comment by lewtun** [comment]
I think this is good to go, so gently pinging @lvwerra @younesbelkada for a final review 🙏

**Review by younesbelkada — APPROVED** [review]
Thanks a lot @lewtun ! This looks great to me, I just have one suggestion about the name of the class, what `RewardTrainingArguments` ? I am also happy with the current naming if @lvwerra agrees

File: trl/trainer/reward_trainer.py [hunk 4]
```diff
@@ -122,11 +126,20 @@ def __init__(
...
                 max_length = 512
+            elif args.max_length is None:
+                warnings.warn(
+                    "When using RewardDataCollatorWithPadding, you should set `max_length` in RewardConfig."
+                    " It will be set to `512` by default, but you should do it yourself in the future.",
+                    UserWarning,
+                )
+                max_length = 512
+            else:
+                max_length = args.max_length
```
Comment (lvwerra) [inline trl/trainer/reward_trainer.py:142]: I think in the case where `max_length` is not `None` then it is still overwritten when `args.max_length is None`, right? I think in that case we should keep the original value to be backwards compatible.

Comment (lewtun) [reply trl/trainer/reward_trainer.py:142]: Oh yes, good catch - I'll fix that

**Comment by lewtun** [comment]
> Thanks a lot @lewtun ! This looks great to me, I just have one suggestion about the name of the class, what `RewardTrainingArguments` ? I am also happy with the current naming if @lvwerra agrees

Are you suggesting we use `RewardTrainingArguments` instead of `RewardConfig`? I originally had the former, but switched to the latter to be aligned with `PPOConfig`. I don't have a strong opinion, but whatever choice we make should stay consistent for all the other trainers, e.g. `SFTTrainingArguments` vs `SFTConfig` etc.

One advantage of `XConfig` classes is that it's less keystrokes :D

File: trl/trainer/reward_trainer.py (hunk 4 above)
Comment (lewtun) [reply trl/trainer/reward_trainer.py]: Fixed in https://github.com/huggingface/trl/pull/726/commits/5b67e0d40d96811b3202ea5bf11628698c44b49b

**Review by lvwerra — APPROVED** [review]
LGTM! 🚀

**Review by younesbelkada — APPROVED** [review]
Thanks @lewtun , makes sense, let's merge it! 🚀
